yt_dlp/extractor/huya.py

```diff
@@ -0,0 +1,140 @@
+    IE_NAME = 'huya:live'
+    IE_DESC = 'huya.com'
+    TESTS = [{
+        'url': 'https://www.huya.com/572329',
+        'info_dict': {
+            'id': '572329',
+            'title': 're:^.+$',
+            'description': 're:^.+$',
+            'is_live': True,
+            'view_count': int,
+        },
+        'params': {
+            'skip_download': True,
+        },
+    }, {
+        'url': 'https://www.huya.com/xiaoyugame',
+        'only_matching': True
+    }]
+
+    _RESOLUTION = {
+        '蓝光4M': {
+            'width': 1920,
+            'height': 1080,
+        },
+        '超清': {
+            'width': 1280,
+            'height': 720,
+        },
+        '流畅': {
+            'width': 800,
+            'height': 480
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id=video_id)
+        json_stream = self._search_regex(r'"stream":\s+"([a-zA-Z0-9+=/]+)"', webpage, 'stream', default=None)
+        if not json_stream:
+            raise ExtractorError('Video is offline', expected=True)
+        stream_data = self._parse_json(compat_b64decode(json_stream).decode(), video_id=video_id,
+                                       transform_source=js_to_json)
+        room_info = try_get(stream_data, lambda x: x['data'][0]['gameLiveInfo'])
+        if not room_info:
+            raise ExtractorError('Can not extract the room info', expected=True)
+        title = room_info.get('roomName') or room_info.get('introduction') or self._html_search_regex(
+            r'<title>([^<]+)</title>', webpage, 'title')
+        screen_type = room_info.get('screenType')
+        live_source_type = room_info.get('liveSourceType')
+        stream_info_list = try_get(stream_data, lambda x: x['data'][0]['gameStreamInfoList'])
+        if not stream_info_list:
+            raise ExtractorError('Can not extract media info.', expected=True)
+        formats = []
+        for stream_info in stream_info_list:
+            stream_url = stream_info.get('sFlvUrl')
+            if not stream_url:
+                continue
+            stream_name = stream_info.get('sStreamName')
+            re_secret = not screen_type and live_source_type in (0, 8, 13)
+            params = dict(compat_urlparse.parse_qsl(unescape(stream_info['sFlvAntiCode'])))
+            fm, ss = '', ''
+            if re_secret:
+                fm, ss = self.encrypt(params, stream_info, stream_name)
+            for si in stream_data.get('vMultiStreamInfo'):
+                rate = si.get('iBitRate')
+                if rate:
+                    params['ratio'] = rate
+                else:
+                    params.pop('ratio', None)
+                if re_secret:
+                    params['wsSecret'] = hashlib.md5(
+                        '_'.join([fm, params.get('u'), stream_name, ss, params.get('wsTime')]))
```

Comment:
```suggestion
                        '_'.join([fm, params['u'], stream_name, ss, params['wsTime']]))
```
or have a fallback value, as having None (default fallback for `dict.get`) will cause `str.join` to error